Implement open inbox for dev-connect

mariocsee opened the pull request:
## What type of PR is this? (check all applicable)
- [x] Feature

## Description
This PR adds open inbox to users that want to allow others to message without mutually following each other. 

The column `inbox_type` is added to Users with "private" as the default value. The only other option right now is "open." This can be changed under `settings/misc`. 

An `inbox_type` of "private" retains the previous (default) behavior of DEV Connect, where only users that mutually follow each other can send messages to one another. Now, if you view a user you follow and that follows you, a chat button will be displayed beside the following button. 

An `inbox_type` of "open" displays a chat button on your profile and allows any user, whether you mutually follow each other or not, to open a chat with you on DEV Connect. This works by calling the `open_chat` method in `users_controller` and creating a chat channel (if none already exist between you and the user with an open inbox) and redirect you to `dev.to/connect/@username`. 

### To-do:
- [x] Add initial message (when a new channel is opened, the default message on top is "You and @mozellmrsprohaska are connected because...", but with an "open" `inbox_type` that message is false and needs to be changed for clarity)
- [x] Chats that are made inactive because of unfollows can be made active again if inbox_type is open
- [x] Rspec tests for `follows_controller` pass given changes in `show` 

## Related Tickets & Documents
resolves #1460 

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)
### Changing your inbox type in `settings/misc`
![settings-change-inboxtype](https://user-images.githubusercontent.com/13403332/53375034-2690a700-3928-11e9-88bb-b2b44fccb7bb.gif)

### Follow Back button for users that follow you
![dev-follow-back](https://user-images.githubusercontent.com/13403332/53375063-35775980-3928-11e9-890c-6eb5ae3cd755.gif)

### Users that you follow and follow you have a chat button that redirects to the chat channel between you
![mutual-follow-chat](https://user-images.githubusercontent.com/13403332/53375138-62c40780-3928-11e9-9c6f-1f9dce0c1324.gif)

### Manually changing inbox type to "open" for testing in local environment
![change-inbox-type](https://user-images.githubusercontent.com/13403332/53375314-e7af2100-3928-11e9-9f04-ceda504cf86f.gif)

### New chat for users with open inbox type
![open-inbox-new-msg](https://user-images.githubusercontent.com/13403332/53375332-f1388900-3928-11e9-8d47-cb33ee964209.gif)

### If you mutually follow someone who has an open inbox, it will prioritize redirecting straight to `connect/@username`

## Added to documentation?

- [x] no documentation needed

benhalpern commented:
Looks good so far. A future thoughts/headsup:

- The "Message Me" button on the profile should exist for open inboxers _and_ folks who have message permission based on mutual follows. Since pages can be edge cached and we can't calculate that on the fly, we will need to do this with JS similar to how we do the follow button. (But I think perhaps we can merge other functionality and figure out the UI stuff later.
- Perhaps instead of boolean `inbox` we have a string `inbox_type` because there could be different types other than closed and opened in the future.
- I'm definitely interested to see how you manage the creation of Chat channels on the fly here. We'll need to make sure it's all logically compatible with mutual-follow inboxes.
- Looks like there are some schema issues with some extra files included here. Not a big deal, but sometimes you need to drop old DB stuff and check for this. I make this mistake too sometimes.

Good luck, this will be awesome!

mariocsee commented:
@benhalpern I just updated the PR description and screenshots. There's some `to-dos` I added as well. The button is implemented in `initializeAllChatButtons.js` which closely follows `initializeAllFollowButtons.js` (but removes a lot of lines that aren't used for chat). Also CC has a lot of issues with the new file that occur in almost all initialize JS files, so I committed with `no-verify` to check everything is working properly first.

In `views/users/show.html.erb`, the chat button is patterned after the follow button but with more `data-info`—including `"userFollowing": <%=@user.follows.map{|x| x.followable_id} %>`. This is used to cross check that both users follow each other (if the `inbox_type` isn't "open").

mariocsee commented:
Refactored how mutual checks are done by extending `show` in `follows_controller.rb` and making use of `FollowChecker` in the opposite direction rather than passing the list of IDs a user follows and searching through that list (since the list can become very large depending on the user). 

Running into issues when: 
- you have an existing chat with User1 you follow and follows you
- you unfollow User1, rendering that chat inactive and hiding it from connect
- User1 happens to have an `"inbox_type"` of "open" so you try chatting with User1 by clicking on the chat button in their profile
- you are redirected to `dev.to/connect/@User1`, but run into an error in console and are in connect with an empty chat

Working on this right now

maestromac commented on app/labor/markdown_parser.rb:
````diff
@@ -124,6 +124,8 @@ def escape_liquid_tags_in_codeblock(content)
     content.gsub(/`{3}.*?`{3}|`{2}.+?`{2}|`{1}.+?`{1}/m) do |codeblock|
       if codeblock[0..2] == "```"
         "\n{% raw %}\n" + codeblock + "\n{% endraw %}\n"
+      elsif codeblock[0..1] == "``"
````
This is removed in master.

mariocsee commented on app/assets/javascripts/initializers/initializeAllChatButtons.js:
```diff
@@ -0,0 +1,94 @@
+
+function initializeChatButton(button) {
+  // if user logged out, do nothing
+  var userStatus = document
```
UserData() returns null when not signed in so will do that instead of checking body for user status

mariocsee commented on app/views/users/show.html.erb:
```diff
@@ -26,6 +26,9 @@
             <button id="user-follow-butt" class="cta follow-action-button user-profile-follow-button" style="color:<%= user_colors(@user)[:text] %>;background-color:<%=user_colors(@user)[:bg] %>" data-info='{"id":<%= @user.id%>,"className":"<%= @user.class.name %>"}'>&nbsp;</button>
           </span>
+          <span class="user-profile-chat-button-wrapper">
+            <button id="user-chat-button" class="cta chat-action-button user-profile-chat-button" style="color:<%= user_colors(@user)[:text] %>;background-color:<%=user_colors(@user)[:bg] %>; display:none" data-info='{"id":<%= @user.id%>,"className":"<%= @user.class.name %>","username":"<%= @user.username %>", "showChat":"<%= @user.inbox_type %>"}'>&nbsp;</button>
```
Put Chat in the button and removed JS code that fills text and styling.

mariocsee commented on app/controllers/users_controller.rb:
```diff
@@ -142,6 +142,35 @@ def follow_hiring_tag(user)
+    common_direct_chats = (visitor.chat_channels & user.chat_channels).select { |channel| channel.channel_type == "direct" }
+
+    if common_direct_chats.empty?
+      chat = ChatChannel.create_with_users([visitor, user], "direct")
```
added if statement that double checks open inbox and remove inefficient common channels check

maestromac commented on app/controllers/chat_channels_controller.rb:
```diff
@@ -93,6 +93,24 @@ def moderate
+      # get message param to generate message to send
+      # message_markdown = "Hi #{chat_recipient.username}! I am #{current_user.username}. I can message you on DEV Connect because your inbox is open."
+      message_markdown = params[:message]
+      puts message_markdown
```
```suggestion
```
Lets move this put statement
